spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

```diff
@@ -213,7 +223,19 @@ class OptimizeExecutor(
   /** Timestamp to use in [[FileAction]] */
   private val operationTimestamp = new SystemClock().getTimeMillis()
 
-  private val isMultiDimClustering = zOrderByColumns.nonEmpty
+  private val isClusteredTable = ClusteredTableUtils.isSupported(txn.snapshot.protocol)
+
+  private val isMultiDimClustering = isClusteredTable || zOrderByColumns.nonEmpty
+
+  private val (clusteringColumns, curve): (Seq[String], String) = {
+    if (zOrderByColumns.nonEmpty) {
+      (zOrderByColumns, "zorder")
+    } else if (isClusteredTable) {
+      (ClusteringColumnInfo.extractLogicalNames(txn.snapshot), "hilbert")
+    } else {
+      Nil -> ""
```

Comment:
is this a valid scenario? when does this happen?

Reply:
This path gets triggered when we call OPTIMIZE on a non-clustered table, so it triggered compaction.

I can't do the assert in https://github.com/delta-io/delta/pull/2461/files#r1446787772 because OPTIMIZE(compaction) will trigger this path and will hit the assert.

Reply:
make it lazy val?

Reply:
Looks like I can't make `clusteringColumns` a lazy val since `getOperation` relies on it being empty for compaction scenario.

So I made `curve` a lazy val and put the assert there.